File: sow-dist/deploy/android/app/src/main/java/com/shadowsofwar/TwaLauncherActivity.java
package com.shadowsofwar;

import android.graphics.Color;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.Window;

import androidx.browser.customtabs.CustomTabsCallback;
import androidx.browser.customtabs.CustomTabsSession;

import com.google.androidbrowserhelper.trusted.LauncherActivity;
import com.google.androidbrowserhelper.trusted.TwaLauncher;
import com.google.android.gms.games.GamesSignInClient;
import com.google.android.gms.games.PlayGames;
import com.google.android.gms.games.PlayGamesSdk;

import org.json.JSONObject;
import org.json.JSONArray;

import java.lang.reflect.Field;
import java.util.UUID;

/** Opens the TWA; the web loader starts optional Play Games authentication. */
public final class TwaLauncherActivity extends LauncherActivity {
    private static final String TAG = "SOW_PGS";
    private static final Uri SOW_ORIGIN = Uri.parse("https://shadowsofwar.io");
    private static volatile CustomTabsSession messageSession;
    private GamesSignInClient signInClient;
    private boolean requestInFlight;
    private String rendezvousId;

    @Override
    protected void onCreate(Bundle state) {
        hideSystemBars();
        super.onCreate(state);
        hideSystemBars();
        Log.i(TAG, "TWA launched; starting Play Games in parallel");
        PlayGamesSdk.initialize(getApplicationContext());
        if (!PlayGamesAuth.isAnonymousOptOut(this)) {
            signInClient = PlayGames.getGamesSignInClient(this);
            checkAuthentication();
        } else {
            Log.i(TAG, "automatic Play Games login suppressed by anonymous mode");
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemBars();
        }
    }

    @Override
    protected CustomTabsCallback getCustomTabsCallback() {
        return new CustomTabsCallback() {
            @Override
            public void onNavigationEvent(int navigationEvent, Bundle extras) {
                if (navigationEvent != CustomTabsCallback.NAVIGATION_FINISHED) {
                    return;
                }
                CustomTabsSession session = findMessageSession();
                if (session == null) {
                    Log.w(TAG, "TWA session unavailable; purchase bridge disabled");
                    return;
                }
                messageSession = session;
                boolean requested = session.requestPostMessageChannel(
                        SOW_ORIGIN, SOW_ORIGIN, new Bundle());
                Log.i(TAG, "purchase bridge channel requested=" + requested);
            }

            @Override
            public void onMessageChannelReady(Bundle extras) {
                CustomTabsSession session = findMessageSession();
                if (session == null) {
                    return;
                }
                messageSession = session;
                JSONArray products = new JSONArray();
                for (String product : PurchaseActivity.storeProductIds()) products.put(product);
                JSONObject ready = new JSONObject();
                try {
                    ready.put("type", "sow_bridge_ready");
                    ready.put("bridge_version", 1);
                    ready.put("capabilities", new JSONArray().put("purchase"));
                    ready.put("products", products);
                } catch (Exception error) {
                    Log.w(TAG, "could not build purchase bridge capabilities", error);
                    return;
                }
                int result = session.postMessage(ready.toString(), null);
                Log.i(TAG, "purchase bridge ready result=" + result);
            }

            @Override
            public void onPostMessage(String message, Bundle extras) {
                handleBridgeMessage(message);
            }
        };
    }

    private void handleBridgeMessage(String message) {
        try {
            JSONObject request = new JSONObject(message);
            if (!"purchase".equals(request.optString("type"))) {
                return;
            }
            String productId = request.optString("product_id", "");
            String appUserId = request.optString("app_user_id", "");
            String requestId = request.optString("request_id", "");
            if (!PurchaseActivity.isStoreProduct(productId)
                    || !PurchaseActivity.isPurchaseUserId(appUserId)
                    || !requestId.matches("[A-Za-z0-9_-]{1,80}")) {
                Log.w(TAG, "rejected invalid purchase bridge request");
                postPurchaseResult(requestId, "error", productId);
                return;
            }
            Intent intent = new Intent(this, PurchaseActivity.class).setData(
                    Uri.parse("sow://purchase").buildUpon()
                            .appendQueryParameter("product_id", productId)
                            .appendQueryParameter("app_user_id", appUserId)
                            .appendQueryParameter("request_id", requestId)
                            .build());
            startActivity(intent);
        } catch (Exception error) {
            Log.w(TAG, "invalid purchase bridge message", error);
        }
    }

    static void postPurchaseResult(String requestId, String status, String productId) {
        CustomTabsSession session = messageSession;
        if (session == null) {
            Log.w(TAG, "purchase result dropped; TWA message session unavailable");
            return;
        }
        try {
            JSONObject result = new JSONObject()
                    .put("type", "purchase_result")
                    .put("request_id", requestId == null ? "" : requestId)
                    .put("status", status);
            if (productId != null) {
                result.put("product_id", productId);
            }
            int code = session.postMessage(result.toString(), null);
            Log.i(TAG, "purchase result posted status=" + status + " code=" + code);
        } catch (Exception error) {
            Log.w(TAG, "could not post purchase result", error);
        }
    }

    private CustomTabsSession findMessageSession() {
        try {
            Field launcherField = LauncherActivity.class.getDeclaredField("mTwaLauncher");
            launcherField.setAccessible(true);
            Object launcher = launcherField.get(this);
            if (launcher == null) {
                return null;
            }
            Field sessionField = TwaLauncher.class.getDeclaredField("mSession");
            sessionField.setAccessible(true);
            return (CustomTabsSession) sessionField.get(launcher);
        } catch (ReflectiveOperationException | ClassCastException error) {
            Log.w(TAG, "could not access TWA session for bridge", error);
            return null;
        }
    }

    private void hideSystemBars() {
        Window window = getWindow();
        window.setStatusBarColor(Color.TRANSPARENT);
        window.setNavigationBarColor(Color.TRANSPARENT);
        window.getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
    }

    @Override
    protected Uri getLaunchingUrl() {
        rendezvousId = UUID.randomUUID().toString().replace("-", "");
        Uri.Builder builder = Uri.parse("https://shadowsofwar.io/play/")
                .buildUpon()
                .appendQueryParameter("sow_platform", "android")
                .appendQueryParameter("sow_playgames_rendezvous", rendezvousId);
        if (PlayGamesAuth.isAnonymousOptOut(this)) {
            builder.appendQueryParameter("sow_playgames_mode", "anonymous");
        }
        return builder.build();
    }

    private void checkAuthentication() {
        if (requestInFlight || signInClient == null) {
            return;
        }
        requestInFlight = true;
        signInClient.isAuthenticated().addOnCompleteListener(task -> {
            Log.i(TAG, "isAuthenticated success=" + task.isSuccessful());
            if (task.isSuccessful() && task.getResult() != null && task.getResult().isAuthenticated()) {
                Log.i(TAG, "automatic Play Games session authenticated");
                PlayGamesAuth.requestServerAccess(
                        this,
                        signInClient,
                        rendezvousId,
                        success -> requestInFlight = false);
            } else {
                requestInFlight = false;
                Log.i(TAG, "automatic Play Games authentication unavailable; continuing anonymously");
            }
        });
    }

}

File: sow-dist/deploy/android/app/src/main/kotlin/com/shadowsofwar/TwaLauncherActivity.kt
package com.shadowsofwar

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Window
import androidx.browser.customtabs.CustomTabsCallback
import androidx.browser.customtabs.CustomTabsSession
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.google.android.gms.games.PlayGames
import com.google.android.gms.games.PlayGamesSdk
import com.google.androidbrowserhelper.trusted.LauncherActivity
import com.google.androidbrowserhelper.trusted.TwaLauncher
import org.json.JSONArray
import org.json.JSONObject
import java.lang.reflect.Field
import java.util.UUID

/** Opens the TWA. Play Games auth starts only after the web loader is ready. */
class TwaLauncherActivity : LauncherActivity() {
    companion object {
        private const val TAG = "SOW_PGS"
        private val SOW_ORIGIN: Uri = Uri.parse("https://shadowsofwar.io")

        @Volatile
        private var messageSession: CustomTabsSession? = null

        @JvmStatic
        fun postPurchaseResult(requestId: String?, status: String, productId: String?) {
            val session = messageSession
            if (session == null) {
                Log.w(TAG, "purchase result dropped; TWA message session unavailable")
                return
            }
            try {
                val result = JSONObject()
                    .put("type", "purchase_result")
                    .put("request_id", requestId.orEmpty())
                    .put("status", status)
                if (productId != null) result.put("product_id", productId)
                val code = session.postMessage(result.toString(), null)
                Log.i(TAG, "purchase result posted status=$status code=$code")
            } catch (error: Exception) {
                Log.w(TAG, "could not post purchase result", error)
            }
        }

        @JvmStatic
        fun postPlayGamesAuthResult(status: String, rendezvousId: String?) {
            val session = messageSession
            if (session == null) {
                Log.w(TAG, "Play Games auth result dropped; TWA message session unavailable")
                return
            }
            try {
                val result = JSONObject()
                    .put("type", "playgames_auth_result")
                    .put("status", status)
                    .put("rendezvous_id", rendezvousId.orEmpty())
                val code = session.postMessage(result.toString(), null)
                Log.i(TAG, "Play Games auth result posted status=$status code=$code")
            } catch (error: Exception) {
                Log.w(TAG, "could not post Play Games auth result", error)
            }
        }
    }

    private var rendezvousId: String? = null
    private var silentAuthInFlight = false

    override fun onCreate(state: Bundle?) {
        hideSystemBars()
        super.onCreate(state)
        hideSystemBars()
        Log.i(
            TAG,
            "artifact source_sha=${BuildConfig.SOW_SOURCE_SHA} " +
                "version_code=${BuildConfig.VERSION_CODE} package=${packageName}"
        )
        Log.i(TAG, "TWA launcher ready; Play Games waits for loader-ready")
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) hideSystemBars()
    }

    override fun getCustomTabsCallback(): CustomTabsCallback = object : CustomTabsCallback() {
        override fun onNavigationEvent(navigationEvent: Int, extras: Bundle?) {
            if (navigationEvent != CustomTabsCallback.NAVIGATION_FINISHED) return
            val session = findMessageSession()
            if (session == null) {
                Log.w(TAG, "TWA session unavailable; native bridge disabled")
                return
            }
            messageSession = session
            val requested = session.requestPostMessageChannel(SOW_ORIGIN, SOW_ORIGIN, Bundle())
            Log.i(TAG, "native bridge channel requested=$requested")
        }

        override fun onMessageChannelReady(extras: Bundle?) {
            val session = findMessageSession() ?: return
            messageSession = session
            val products = JSONArray()
            PurchaseActivity.storeProductIds().forEach(products::put)
            val ready = JSONObject()
                .put("type", "sow_bridge_ready")
                .put("bridge_version", 2)
                .put(
                    "capabilities",
                    JSONArray()
                        .put("purchase")
                        .put("restore")
                        .put("playgames_silent_auth")
                )
                .put("products", products)
            val result = session.postMessage(ready.toString(), null)
            Log.i(TAG, "native bridge ready result=$result")
        }

        override fun onPostMessage(message: String, extras: Bundle?) {
            handleBridgeMessage(message)
        }
    }

    private fun handleBridgeMessage(message: String) {
        try {
            val request = JSONObject(message)
            when (request.optString("type")) {
                "playgames_silent_auth" -> startSilentPlayGamesAuth(request)
                "purchase" -> handlePurchaseRequest(request)
                "restore" -> handleRestoreRequest(request)
            }
        } catch (error: Exception) {
            Log.w(TAG, "invalid native bridge message", error)
        }
    }

    private fun startSilentPlayGamesAuth(request: JSONObject) {
        if (silentAuthInFlight) return
        val localRendezvous = rendezvousId
        val requestedRendezvous = request.optString("rendezvous_id", "")
        Log.i(TAG, "loader-ready Play Games silent auth requested")
        if (localRendezvous.isNullOrBlank() || requestedRendezvous != localRendezvous) {
            postPlayGamesAuthResult("error", localRendezvous)
            return
        }
        if (PlayGamesAuth.isAnonymousOptOut(this)) {
            postPlayGamesAuthResult("unavailable", localRendezvous)
            return
        }

        silentAuthInFlight = true
        PlayGamesSdk.initialize(applicationContext)
        val client = PlayGames.getGamesSignInClient(this)
        client.isAuthenticated().addOnCompleteListener { task ->
            val authenticated = task.isSuccessful && task.getResult()?.isAuthenticated == true
            Log.i(TAG, "loader-ready isAuthenticated success=${task.isSuccessful} authenticated=$authenticated")
            if (!authenticated) {
                silentAuthInFlight = false
                postPlayGamesAuthResult("unavailable", localRendezvous)
                return@addOnCompleteListener
            }
            PlayGamesAuth.requestServerAccess(this, client, localRendezvous) { success ->
                silentAuthInFlight = false
                if (success) PlayGamesAuth.setAnonymousOptOut(this, false)
                postPlayGamesAuthResult(if (success) "ready" else "error", localRendezvous)
            }
        }
    }

    private fun handlePurchaseRequest(request: JSONObject) {
        val productId = request.optString("product_id", "")
        val appUserId = request.optString("app_user_id", "")
        val requestId = request.optString("request_id", "")
        if (!PurchaseActivity.isStoreProduct(productId) ||
            !PurchaseActivity.isPurchaseUserId(appUserId) ||
            !validRequestId(requestId)
        ) {
            Log.w(TAG, "rejected invalid purchase bridge request")
            postPurchaseResult(requestId, "error", productId)
            return
        }
        val data = Uri.parse("sow://purchase").buildUpon()
            .appendQueryParameter("product_id", productId)
            .appendQueryParameter("app_user_id", appUserId)
            .appendQueryParameter("request_id", requestId)
            .build()
        startActivity(Intent(this, PurchaseActivity::class.java).setData(data))
    }

    private fun handleRestoreRequest(request: JSONObject) {
        val appUserId = request.optString("app_user_id", "")
        val requestId = request.optString("request_id", "")
        if (!PurchaseActivity.isPurchaseUserId(appUserId) || !validRequestId(requestId)) {
            Log.w(TAG, "rejected invalid restore bridge request")
            postPurchaseResult(requestId, "error", null)
            return
        }
        val data = Uri.parse("sow://restore").buildUpon()
            .appendQueryParameter("app_user_id", appUserId)
            .appendQueryParameter("request_id", requestId)
            .build()
        startActivity(Intent(this, PurchaseActivity::class.java).setData(data))
    }

    private fun validRequestId(value: String): Boolean = value.matches(Regex("[A-Za-z0-9_-]{1,80}"))

    private fun findMessageSession(): CustomTabsSession? {
        return try {
            val launcherField: Field = LauncherActivity::class.java.getDeclaredField("mTwaLauncher")
            launcherField.isAccessible = true
            val launcher = launcherField.get(this) ?: return null
            val sessionField: Field = TwaLauncher::class.java.getDeclaredField("mSession")
            sessionField.isAccessible = true
            sessionField.get(launcher) as? CustomTabsSession
        } catch (error: ReflectiveOperationException) {
            Log.w(TAG, "could not access TWA session for bridge", error)
            null
        } catch (error: ClassCastException) {
            Log.w(TAG, "TWA session has an unexpected type", error)
            null
        }
    }

    private fun hideSystemBars() {
        val currentWindow: Window = window
        WindowCompat.setDecorFitsSystemWindows(currentWindow, false)
        WindowInsetsControllerCompat(currentWindow, currentWindow.decorView).apply {
            hide(WindowInsetsCompat.Type.systemBars())
            systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        }
    }

    override fun getLaunchingUrl(): Uri {
        rendezvousId = UUID.randomUUID().toString().replace("-", "")
        val builder = Uri.parse("https://shadowsofwar.io/play/").buildUpon()
            .appendQueryParameter("sow_platform", "android")
            .appendQueryParameter("sow_playgames_rendezvous", rendezvousId)
            .appendQueryParameter("sow_source_sha", BuildConfig.SOW_SOURCE_SHA)
        if (PlayGamesAuth.isAnonymousOptOut(this)) {
            builder.appendQueryParameter("sow_playgames_mode", "anonymous")
        }
        return builder.build()
    }
}
